src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/PackagesConfigToPackageReferenceMigrator.cs

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Common;
+using NuGet.PackageManagement.Telemetry;
+using NuGet.Packaging.Core;
+using NuGet.ProjectManagement;
+using NuGet.ProjectManagement.Projects;
+using NuGet.Protocol.Core.Types;
+
+namespace NuGet.PackageManagement.UI
+{
+    internal static class PackagesConfigToPackageReferenceMigrator
+    {
+        internal static async Task<string> DoUpgradeAsync(
+            INuGetUIContext context,
+            INuGetUI uiService,
+            NuGetProject nuGetProject,
+            IEnumerable<NuGetProjectUpgradeDependencyItem> upgradeDependencyItems,
+            IEnumerable<PackageIdentity> notFoundPackages,
+            bool collapseDependencies,
+            IProgress<ProgressDialogData> progress,
+            CancellationToken token)
+        {
+
+            var startTime = DateTimeOffset.Now;
+            var packagesCount = 0;
+            var status = NuGetOperationStatus.Succeeded;
+
+
+            using (var telemetry = new TelemetryActivity(Guid.Empty))
+            {
+                try
+                {
+                    // 1. Backup files (csproj and packages.config) that will change
```

Comment:
nit: 
For better readibility, consider extracting these steps in separate methods each. 
Otherwise, we get NuGetPackageManager V2 :)